### DISCUSSION TITLE
How do I implement drawing on the react-flow canvas?

### DISCUSSION THREAD

**Discussion by ramseskamanda** [xyflow/xyflow#1492] [Q&A]
Hi everyone!

I'm building a collaborative tool using this package, and I'm trying to add drawing functionality to the canvas. Would anyone be able to set me on the right path? I've already built all the drawing logic on a separate canvas that would be positioned on top/as a child of ReactFlow (similar to how the Background component is implemented), however, allowing both drawing events and panning events at the same time is proving to be more difficult than I anticipated. Would anyone know how to implement something like that?

If anyone needs an example of what I'm trying to implement, it would be like Figma, where one can draw anywhere on an infinite canvas and interact with the canvas at the same time through panning.

Thank you in advance for your help!

**Comment by datoslabs** [comment]
Interesting problem!  At a high level, you would need to (at a minimum):

1.  Synchronize pan/zoom between react-flow and canvas, and
2.  Separate pan/zoom control from drawing.

For #1, have you tried using the `onMove` event handler to retrieve latest zoom level as well as translation information via store state hook - `const [x,y, zoom] = useStoreState((state) => state.transform);` and applying the same to canvas' 2D ctx?  I am not sure how well the zoom level and x/y coordinates translate to canvas 2D ctx's scale and transform matrix, you may need to experiment a little bit so the canvas overlays correctly against react-flow's graph after pan/zoon.

As for #2, there are many ways to separate the 2 concerns, a potential solution may be to use hot-keys + mouse movements to enable pan/zoom.

**Comment by Th3R1c0** [comment]
lock the canvas and ability to select other nodes when drawing, the users drawing is simply a node that changes width and height as the user draws
